lib/membrane_matroska/muxer.ex

```diff
@@ -0,0 +1,396 @@
+defmodule Membrane.Matroska.Muxer do
+  @moduledoc """
+  Filter element for muxing Matroska files.
+
+  It accepts an arbitrary number of Opus, VP8, VP9 or H264 streams and outputs a bytestream in Matroska format containing those streams.
+
+  Muxer guidelines
+  https://www.matroskaproject.org/docs/container/
+  """
+  use Bitwise
+  use Bunch
+
+  use Membrane.Filter
+
+  alias Membrane.{Buffer, RemoteStream}
+  alias Membrane.{Opus, VP8, VP9, MP4}
+  alias Membrane.Matroska.Parser.Codecs
+  alias Membrane.Matroska.Serializer
+  alias Membrane.Matroska.Serializer.Helper
+  alias Membrane.MP4.Payload.AVC1
+
+  def_options title: [
+                spec: String.t(),
+                default: "Membrane Matroska file",
+                description: "Title to be used in the `Segment/Info/Title` element"
+              ],
+              date: [
+                spec: nil | integer(),
+                default: nil,
+                description: "Datetime which will be store in  the `Segment/Info/DateUTC` element.
+                  Default value is the time of initialization of this element."
+              ]
+
+  # tags:
+
+  def_input_pad :input,
+    availability: :on_request,
+    mode: :pull,
+    demand_unit: :buffers,
+    caps: [
+      Opus,
+      {RemoteStream, content_format: Membrane.Caps.Matcher.one_of([VP8, VP9]), type: :packetized},
+      MP4.Payload
+      # Why doesn't match correctly {MP4.Payload, content: AVC1}
+    ]
+
+  def_output_pad :output,
+    availability: :always,
+    mode: :pull,
+    caps: {RemoteStream, content_format: :Matroska}
+
+  # 5 mb
+  @cluster_bytes_limit 5_242_880
+  @cluster_time_limit Membrane.Time.seconds(5)
+  @timestamp_scale Membrane.Time.millisecond()
+
+  defmodule State do
+    @moduledoc false
+    defstruct tracks: %{},
+              segment_position: 0,
+              expected_tracks: 0,
+              active_tracks: 0,
+              cluster_acc: Helper.serialize({:Timecode, 0}),
+              cluster_time: nil,
+              cluster_size: 0,
+              cues: [],
+              time_min: 0,
+              time_max: 0,
+              options: %{}
+  end
+
+  @impl true
+  def handle_init(options) do
+    options = options |> Map.put(:duration, 0) |> Map.put(:clusters_size, 0)
+    {:ok, %State{options: options}}
+  end
+
+  @impl true
+  def handle_pad_added(_pad, context, state) when context.playback_state != :playing do
+    {:ok, %State{state | expected_tracks: state.expected_tracks + 1}}
+  end
+
+  @impl true
+  def handle_pad_added(_pad, _context, _state) do
+    raise "Can't add new input pads to muxer in state :playing!"
+  end
+
+  @impl true
+  def handle_caps(Pad.ref(:input, id), _caps, _context, state)
+      when is_map_key(state.tracks, id) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(Pad.ref(:input, id), caps, _context, state) do
```

Comment:
Ok, a little problem here. I assume a reference won't cut it as id? If you don't use `via_in` in link builder, we will get a ref here. We need to either raise or better, generate random id in this case. I think you should be able to get this random generation sorted relatively quickly, just don't assume that pad id is equal to Matrioska id and get it from `state.tracks[pad_id].id` instead